docs: Added Thanos Go style guide and some development tips. (#2359)
thanos-io/thanos: CONTRIBUTING.md

* Documentation
* Code comments

## Development

> Following section explains various suggestions and procedures while development of Thanos.

Reviewer: The following section..

Reviewer: .. and procedures to note during development of Thanos.

### Prerequisites

> * It is strongly recommended that you use OSX or popular Linux distributions systems e.g. Ubuntu, Redhat, or OpenSUSE for development.

Reviewer: Why is it strongly recommended to use a `popular` linux distro vs something like Nix or Gentoo?

Reviewer: s/Redhat/Red Hat/

> * Go SDK 1.13.9 or newer installed.

Reviewer 2: ```suggestion
* Go 1.13.9 or newer installed.
```

> ### First steps

Reviewer: s/steps/Steps/

> It's a key to get familiar with style guides and mechanics of Thanos, especially if your contribution touches more than on component of the Thanos distributed system.

Reviewer: ```suggestion
It's key to get familiarized with style guide and mechanics of Thanos, especially if your contribution touches more than one component of the Thanos distributed system.
```

We recommend:

> * Reading [getting started docs](docs/getting-started.md) and getting through this. Alternatively https://katacoda.com/thanos.

Reviewer: ```suggestion
* Reading the [getting started docs](docs/getting-started.md) and working through them, or alternatively working through the [Thanos tutorials](https://katacoda.com/thanos).
```

> * Familiarize yourself with our [coding style guides.](docs/contributing/coding-style-guide.md).

Reviewer: grammar, to keep this list correct, we need the start of each item to end in `-ing`. Otherwise one is an action while another is an instruction.

```suggestion
* Familiarizing yourself with our [coding style guidelines.](docs/contributing/coding-style-guide.md).
```

> * Familiarize yourself with [Makefile](Makefile) commands like `format`, `build`, `proto`, `docker` and `test`. `make help` can print you some

Reviewer: ```suggestion
* Familiarizing yourself with the [Makefile](Makefile) commands, for example `format`, `build`, `proto`, `docker` and `test`. `make help` will print some
```

> available commands with explanation.

Reviewer: ```suggestion
available commands with details.
```

$ ./thanos -h
```

> 1. Keep PRs as small as possible. For each of your PR, you create one branch based on the latest master. Chain them if needed (base PR on other PRs). Here are sample steps you can follow. You can get more details about the workflow from [here](https://gist.github.com/Chaser324/ce0505fbed06b947d962).

Reviewer: ```suggestion
1. Keep PRs as small as possible. For each of your PRs, create a new branch based on the latest master. Chain them if needed (base one PR on other PRs). You can read more details about the workflow [here](https://gist.github.com/Chaser324/ce0505fbed06b947d962).
```

$ git push origin <your_branch_for_new_pr>
```

> 1. Add unit tests for new functionality. Add e2e tests if functionality is major.

Reviewer: ```suggestion
1. Add unit tests for new functionality. Add e2e tests for major changes to functionality.
```

> 1. If you don't have a live object store ready you can use `make test-local` command.

Reviewer: ```suggestion
1. If you don't have a live object store ready, you can use the `make test-local` command.
```

> NOTE: This command skips live object storages by specifying environment variables, so the store specific tests will be run

Reviewer: ```suggestion
NOTE: this command skips tests against live object storage systems by specifying environment variables; this causes the store-specific tests to be run
```

> against memory and filesystem object storages only. The CI runs `make test-ci` instead (uses GCS, AWS).

Reviewer: ```suggestion
against memory and filesystem object storages only. The CI tests run `make test-ci` instead (uses GCS, AWS).
```

> Not having these variables will produce auth errors against GCS, AWS, Azure, COS etc.

Reviewer: ```suggestion
Not specifying these variables will result in auth errors against GCS, AWS, Azure, COS etc.
```

> 1. If your change affects users (adds or removes feature) consider adding the item to [CHANGELOG](CHANGELOG.md)

Reviewer: ```suggestion
1. If your change affects users (adds or removes feature) consider adding the item to the [CHANGELOG](CHANGELOG.md)
```

> 1. You may merge the Pull Request in once you have the sign-off of at least one developers with write access, or if you

Reviewer: ```suggestion
1. You may merge the Pull Request once you have the sign-off of at least one developer with write access, or if you
```

   do not have permission to do that, you may request the second reviewer to merge it for you.

> 1. If you feel like your PR waits too long for a review, feel free to ping [`#thanos-prs`](https://slack.cncf.io/) channel on our slack for review!

Reviewer: ```suggestion
1. If you feel like your PR is waiting too long for a review, feel free to ping the [`#thanos-prs`](https://slack.cncf.io/) channel on our slack for a review!
```

```

You have to commit the changes to `go.mod` and `go.sum` before submitting the pull request.

### Advanced testing

> At some point during development it is useful on top of unit or e2e tests, to run and play with Thanos components manually. While you

Reviewer: ```suggestion
At some point during development it is useful, in addition to running unit or e2e tests, to run and play with Thanos components manually. While you
```

> can run those components by crafting certain flags to some test setup there are already nice tools and scripts available.

Reviewer: ```suggestion
can run the components manually by crafting specific flags for a test setup, there are already some nice tools and scripts available.
```

> Consider following methods:

Reviewer: ```suggestion
Consider the following methods:
```

> * [quickstart.sh](https://github.com/thanos-io/thanos/blob/b08c0ea62abfe4dcf1400da0e37598f0cd8fa8cf/scripts/quickstart.sh): Script that spins

Reviewer: ```suggestion
* [quickstart.sh](https://github.com/thanos-io/thanos/blob/b08c0ea62abfe4dcf1400da0e37598f0cd8fa8cf/scripts/quickstart.sh): this script spins
```

> up example, simple setup. Do `make build` before running to build `thanos` binary first.

Reviewer: ```suggestion
up a simple example setup. Do `make build` before running the script to build the `thanos` binary first.
```

> * `make test-e2e`: The e2e are covering most of the setups and actions we offer with Thanos. It's extremely easy to add `time.Sleep(10* time.Minutes)`

Reviewer: ```suggestion
* `make test-e2e`: the e2e cover most of the setups and functionality Thanos offers. It's extremely easy to add `time.Sleep(10* time.Minutes)`
```

> in certain point of tests (e.g for compactor [here](https://github.com/thanos-io/thanos/blob/8f492a9f073f819019dd9f044e346a1e1fa730bc/test/e2e/compact_test.go#L379)).

Reviewer: ```suggestion
at certain points in the tests (e.g for compactor [here](https://github.com/thanos-io/thanos/blob/8f492a9f073f819019dd9f044e346a1e1fa730bc/test/e2e/compact_test.go#L379)).
```

> This way when you run `make test-e2e` which will sleep for some time, allowing you to connect to the setup manually. Use port that will be printed. For example:

Reviewer: ```suggestion
This way when you run `make test-e2e`, the test will sleep for some time, allowing you to connect to the setup manually using the port printed in the logs. For example:
```

```bash
querier-1: level=info name=querier-1 ts=2020-04-01T12:53:56.101029491Z caller=http.go:56 service=http/server component=query msg="listening for requests and metrics" address=:80
querier-1: level=info name=querier-1 ts=2020-04-01T12:53:56.101106805Z caller=intrumentation.go:48 msg="changing probe status" status=ready
querier-1: level=info name=querier-1 ts=2020-04-01T12:53:56.101290229Z caller=grpc.go:106 service=gRPC/server component=query msg="listening for StoreAPI gRPC" address=:9091
Ports for container: e2e_test_store_gateway-querier-1 Mapping: map[80:32825 9091:32824]
```

> Means that HTTP endpoint will be available on `http://localhost:32825`. You can quickly craft your own test case with our framework as well!

Reviewer: ```suggestion
This output indicates that the HTTP endpoint will be available on `http://localhost:32825`. You can quickly craft your own test case with our framework as well!
```

> NOTE: `make docker` has to work for `make test-e2e`. This currently does not work properly on MacOS.

Reviewer: ```suggestion
NOTE: `make docker` has to work in order for `make test-e2e` to run. This currently does not work properly on MacOS.
```
